### PR TITLE
feat(repository): add datasource method on repository mixin

### PR DESCRIPTION
**Pull request by truongminh**
The PR adds a new dataSource api to `RepositoryMixin` as described in https://github.com/strongloop/loopback-next/issues/743

### Check List
- [x] Add code to allow bindings
- [x] Write tests
- [x] Update examples accordingly (only `example-getting-started`)
- [x] Update TSDocs

### CODE REVIEW

File: packages/repository/src/repository-mixin.ts [hunk 1]
```diff
@@ -71,6 +72,26 @@ export function RepositoryMixin<T extends Class<any>>(superClass: T) {
...
+     * const ds: juggler.DataSource = new DataSourceConstructor({
+     *   name: 'db',
+     *   connector: 'memory',
+     * });
+     *
+     * app.dataSource(ds);
+     * ```
+     */
+    dataSource(datasource: juggler.DataSource) {
+      const datasourceKey = `datasource`;
```
Comment (bajtos) [inline packages/repository/src/repository-mixin.ts]: Typically, an application can have multiple datasources. We should build a unique binding key for each datasource, e.g,

```
const datasourceKey = `datasources:${datasource.name}`;
```

Comment (truongminh) [reply packages/repository/src/repository-mixin.ts]: In package `example-getting-started`, the datasource is created dynamically 
```ts
const datasource =
      this.options && this.options.datasource
        ? new DataSourceConstructor(this.options.datasource)
        : db;
```
When it come to the repository constructor, how should we determine the needed datasource?

Comment (truongminh) [reply packages/repository/src/repository-mixin.ts]: For `DataSourceConstructor`
```ts
const ds: juggler.DataSource = new DataSourceConstructor({
      name: 'db',
      connector: 'memory',
    });
```
It turns out that `ds.name` equals to `'memory'` not `'db'`

Comment (bajtos) [reply packages/repository/src/repository-mixin.ts]: > It turns out that ds.name equals to 'memory' not 'db'

This looks like a bug in juggler to me.

I think there should be two ways how to create a named datasource:

```ts
const ds: juggler.DataSource = new DataSourceConstructor('db', {connector: 'memory'});

const ds: juggler.DataSource = new DataSourceConstructor({
  name: 'db',
  connector: 'memory',
});
```

Ideally, both options should return a `DataSource` instance with `ds.name` set to `db`.

Would you like to take a look at this yourself?  I think the changes will be needed in juggler's [lib/datasource.js](https://github.com/strongloop/loopback-datasource-juggler/blob/309b422425a494fea312cf39126d728d7ba4ec46/lib/datasource.js), just remember to add a test verifying your fix (most likely to [test/datasource.js](https://github.com/strongloop/loopback-datasource-juggler/blob/309b422425a494fea312cf39126d728d7ba4ec46/test/datasource.test.js)).

Please mention my GitHub handle in pull request description so that I get a notification to review your changes.

File: packages/repository/src/repository-mixin.ts [hunk 2]
```diff
@@ -71,6 +72,32 @@ export function RepositoryMixin<T extends Class<any>>(superClass: T) {
...
+     * app.dataSource(ds);
+     *
+     * The datasource can be injected with
+     * constructor(@inject('datasource.db') protected datasource: DataSourceType) {
+     *
+     * }
+     * ```
+     */
+    dataSource(dataSource: juggler.DataSource, name?: string) {
+      const dataSourceKey = `datasource.${name || dataSource.name}`;
```
Comment (bajtos) [inline packages/repository/src/repository-mixin.ts]: IIRC, we are using plural form for prefixes, e.g. `controllers.*`. I think we should be using `datasources.*` prefix for consistency.

@raymondfeng thoughts?

File: packages/repository/test/unit/repository-mixin/repository-mixin.test.ts [hunk 3]
```diff
@@ -42,6 +42,39 @@ describe('RepositoryMixin', () => {
...
+      name: 'foo',
+      connector: 'memory',
+    });
+    myApp.dataSource(fooDataSource);
+    expect(myApp.find('datasource.*').map(d => d.key)).to.containEql(
+      'datasource.foo',
+    );
+    expect(myApp.getSync(`datasource.foo`)).to.be.eql(fooDataSource);
+
+    // binding with explicit name
```
Comment (bajtos) [inline packages/repository/test/unit/repository-mixin/repository-mixin.test.ts]: One logical assertion per test please.

The code below should be extracted into a new test.

File: packages/repository/test/unit/repository-mixin/repository-mixin.test.ts [hunk 4]
```diff
@@ -42,6 +42,39 @@ describe('RepositoryMixin', () => {
     expectNoteRepoToBeBound(myApp);
   });
 
+  it('mixed class has .dataSource()', () => {
```
Comment (bajtos) [inline packages/repository/test/unit/repository-mixin/repository-mixin.test.ts]: The source code line defining a test should read like an English sentence. Right now, it does not: "it mixed class has .dataSource()".

I am proposing the following changes:

 - add a new `describe` block to group all tests related to `app.dataSource()` method.
 - rename this test to `it('is mixed into the target class')`

```ts
describe('dataSource()', () => {
  it('is mixed into the target class', () => {
    // ...
  });

  it('binds dataSource using dataSource name for the binding key', () => {
    // ...
  });

  it('binds dataSource using the provided binding key', () => {
    // ...
  });
});
```

File: packages/example-getting-started/config/datasources.json [hunk 5]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "ds",
+  "name": "todo",
```
Comment (bajtos) [inline packages/example-getting-started/config/datasources.json]: Please note that a single datasource can back multiple models, therefore the datasource name is very rarely the same as the model name.

In LoopBack 3.x, we were using `db` as the name of the "main" datasource. Could you please use "db" in this example project too?

Comment (truongminh) [reply packages/example-getting-started/config/datasources.json]: Yes, we should keep it consistent to `db`

File: packages/repository/test/unit/repository-mixin/repository-mixin.test.ts (hunk 4 above)
Comment (truongminh) [reply packages/repository/test/unit/repository-mixin/repository-mixin.test.ts]: I was too lazy.
Let me refactor it.

File: packages/repository/test/unit/repository-mixin/repository-mixin.test.ts (hunk 3 above)
Comment (truongminh) [reply packages/repository/test/unit/repository-mixin/repository-mixin.test.ts]: Sure, I already refactored the block.

File: packages/repository/src/repository-mixin.ts (hunk 2 above)
Comment (truongminh) [reply packages/repository/src/repository-mixin.ts]: Let us keep the style with the prefix `datasources.*`

File: packages/example-getting-started/test/acceptance/application.test.ts [hunk 6]
```diff
@@ -104,6 +104,7 @@ describe('Application', () => {
         port: 0,
       },
       datasource: {
+        name: 'ds',
```
Comment (bajtos) [inline packages/example-getting-started/test/acceptance/application.test.ts]: `db`?

File: packages/repository/test/unit/repository-mixin/repository-mixin.test.ts [hunk 7]
```diff
@@ -128,3 +128,48 @@ describe('RepositoryMixin', () => {
     expect(componentInstance).to.be.instanceOf(component);
   }
 });
+
+describe('RepositoryMixin dataSource', () => {
+  it('is mixed into the target class', () => {
```
Comment (shimks) [inline packages/repository/test/unit/repository-mixin/repository-mixin.test.ts]: grammar mistake; should be `mixes into the target class`

File: packages/repository/test/unit/repository-mixin/repository-mixin.test.ts [hunk 8]
```diff
@@ -128,3 +128,48 @@ describe('RepositoryMixin', () => {
...
+    const myApp = new AppWithRepoMixin();
+    expect(typeof myApp.dataSource).to.be.eql('function');
+  });
+
+  it('is not conflict with previous binding key', () => {
```
Comment (shimks) [inline packages/repository/test/unit/repository-mixin/repository-mixin.test.ts]: `does not conflict ...`

File: packages/repository/test/unit/repository-mixin/repository-mixin.test.ts [hunk 9]
```diff
@@ -128,3 +128,48 @@ describe('RepositoryMixin', () => {
...
+    const myApp = new AppWithRepoMixin();
+    const fooDataSource: juggler.DataSource = new DataSourceConstructor({
+      name: 'foo',
+      connector: 'memory',
+    });
+    myApp.dataSource(fooDataSource);
+    expectDataSourceToBeBound(myApp, fooDataSource, 'foo');
+  });
+
+  it('binds dataSource to a binding key using the name param', () => {
```
Comment (shimks) [inline packages/repository/test/unit/repository-mixin/repository-mixin.test.ts]: nitpick: I think the test name would be phrased with `... using the given argument/name/string` instead of `name param`

File: packages/repository/src/repository-mixin.ts [hunk 10]
```diff
@@ -71,6 +72,32 @@ export function RepositoryMixin<T extends Class<any>>(superClass: T) {
       this.bind(repoKey).toClass(repo);
     }
 
+    /**
+     * Add the dataSource to this application.
+     *
+     * @param dataSource The dataSource to add.
+     * @param name The binding name of the datasource; default to dataSource.name
```
Comment (shimks) [inline packages/repository/src/repository-mixin.ts]: grammar: `defaults to ...`

File: packages/repository/src/repository-mixin.ts [hunk 11]
```diff
@@ -71,6 +72,32 @@ export function RepositoryMixin<T extends Class<any>>(superClass: T) {
...
+     *
+     * ```ts
+     *
+     * const ds: juggler.DataSource = new DataSourceConstructor({
+     *   name: 'db',
+     *   connector: 'memory',
+     * });
+     *
+     * app.dataSource(ds);
+     *
```
Comment (shimks) [inline packages/repository/src/repository-mixin.ts]: We should separate the code blocks here, or comment out precede `the datasource can ...` with //

File: packages/example-getting-started/src/repositories/todo.repository.ts [hunk 12]
```diff
@@ -11,7 +11,7 @@ export class TodoRepository extends DefaultCrudRepository<
   Todo,
   typeof Todo.prototype.id
 > {
-  constructor(@inject('datasource') protected datasource: DataSourceType) {
+  constructor(@inject('datasources.ds') protected datasource: DataSourceType) {
```
Comment (shimks) [inline packages/example-getting-started/src/repositories/todo.repository.ts]: We also need to change the injection key to `datasources.db` if you were to apply @bajtos change

File: packages/repository/src/repository-mixin.ts [hunk 13]
```diff
@@ -71,6 +72,32 @@ export function RepositoryMixin<T extends Class<any>>(superClass: T) {
...
+     * The datasource can be injected with
+     * constructor(@inject('datasource.db') protected datasource: DataSourceType) {
```
Comment (shimks) [inline packages/repository/src/repository-mixin.ts]: `@inject('datasources.db')` (plural)

**Comment by kjdelisle** [comment]
@truongminh Any updates for this pull request? It just needs a few very minor tweaks and then it can land. :D

**Comment by truongminh** [comment]
@kjdelisle yes, there should be some updates as requested in @shimks' review
Let me come back with this work in a couple of hours.

File: packages/example-getting-started/src/application.ts [hunk 14]
```diff
@@ -16,6 +16,7 @@ import {
   Repository,
   DataSourceConstructor,
   RepositoryMixin,
+  juggler,
```
[the remark this answers is not in the document]
Comment (truongminh) [reply packages/example-getting-started/src/application.ts]: Without `juggler` imported, the compiler says `[ts] 'extends' clause of exported class 'TodoApplication' has or is using private name 'juggler.DataSource'.`

File: packages/repository/src/repository-mixin.ts (hunk 11 above)
Comment (truongminh) [reply packages/repository/src/repository-mixin.ts]: The comment is like bellow, isn't it?

File: packages/repository/package.json [hunk 15]
```diff
@@ -26,8 +26,8 @@
     "@loopback/testlab": "^0.1.0"
   },
   "dependencies": {
-    "@loopback/context": "^0.1.0",
-    "loopback-datasource-juggler": "^3.9.2"
+    "@loopback/context": "^4.0.0-alpha.31",
```
Comment (shimks) [inline packages/repository/package.json]: We've recently changed our versioning format, so please revert the change made here.

File: package.json [hunk 16]
```diff
@@ -17,7 +17,9 @@
     "@types/node": "^8.5.9",
     "coveralls": "^3.0.0",
     "cz-conventional-changelog": "^2.1.0",
-    "lerna": "^2.8.0"
+    "lerna": "^2.8.0",
+    "@types/uuid": "^3.4.3",
+    "debug": "^3.1.0"
```
Comment (bajtos) [inline package.json]: I find these two new entries rather suspicious. First of all, they are not sorted alphabetically. Secondly, the top-level `package.json` files should contain only dev-dependencies related to overal project setup. Everything else should go to individual package package.json, e.g. `packages/repository/package.json`.

I'll fix this problem myself to speed up landing of this patch.

**Comment by bajtos** [comment]
Landed 🎉 Thank you for the contribution! ❤️
